plugins/src/gcc/imp.rs

```diff
@@ -0,0 +1,1270 @@
+
+// From 5.5 ""Close" is defined as three standard deviations around this average"
+const STANDARD_DEVIATION_CLOSE_NUM: f64 = 3.;
+
+// Minimal duration between 2 updates on the lost based rate controller
+static LOSS_UPDATE_INTERVAL: Lazy<time::Duration> = Lazy::new(|| time::Duration::from_millis(200));
+static LOSS_DECREASE_THRESHOLD: f64 = 0.1;
+static LOSS_INCREASE_THRESHOLD: f64 = 0.02;
+static LOSS_INCREASE_FACTOR: f64 = 1.05;
+
+// Minimal duration between 2 updates on the lost based rate controller
+static DELAY_UPDATE_INTERVAL: Lazy<time::Duration> = Lazy::new(|| time::Duration::from_millis(100));
+
+static ROUND_TRIP_TIME_WINDOW_SIZE: usize = 100;
+
+fn ts2dur(t: gst::ClockTime) -> Duration {
+    Duration::nanoseconds(t.nseconds() as i64)
+}
+
+fn dur2ts(t: Duration) -> gst::ClockTime {
+    gst::ClockTime::from_nseconds(t.num_nanoseconds().unwrap() as u64)
+}
+
+#[derive(Debug)]
+enum BandwidthEstimationOp {
+    /// Don't update target bitrate
+    Hold,
+    /// Decrease target bitrate
+    Decrease {
+        #[allow(dead_code)]
+        reason: String, // for Debug
+    },
+    Increase,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum ControllerType {
+    // Running the "delay-based controller"
+    Delay,
+    // Running the "loss based controller"
+    Loss,
+}
+
+#[derive(Debug, Clone, Copy)]
+struct Packet {
+    departure: Duration,
+    arrival: Duration,
+    size: usize,
+}
+
+fn human_kbits<T: Into<f64>>(bits: T) -> String {
+    format!("{:.2}kb", (bits.into() / 1_000.))
+}
+
+impl Packet {
+    fn from_structure(structure: &gst::Structure) -> Option<Self> {
+        let lost = structure.get::<bool>("lost").unwrap();
+        let departure = match structure.get::<gst::ClockTime>("local-ts") {
+            Err(e) => {
+                gst::fixme!(
+                    CAT,
+                    "Got packet feedback without local-ts: {:?} - what does that mean?",
+                    e
+                );
+                return None;
+            }
+            Ok(ts) => ts,
+        };
+
+        if lost {
+            return Some(Packet {
+                arrival: Duration::zero(),
+                departure: ts2dur(departure),
+                size: structure.get::<u32>("size").unwrap() as usize,
+            });
+        }
+
+        let arrival = structure.get::<gst::ClockTime>("remote-ts").unwrap();
+
+        Some(Packet {
+            arrival: ts2dur(arrival),
+            departure: ts2dur(departure),
+            size: structure.get::<u32>("size").unwrap() as usize,
+        })
+    }
+}
+
+#[derive(Clone)]
+struct PacketGroup {
+    packets: Vec<Packet>,
+    departure: Duration,       // ms
+    arrival: Option<Duration>, // ms
+}
+
+fn pdur(d: &Duration) -> String {
+    let stdd = time::Duration::from_nanos(d.num_nanoseconds().unwrap().abs() as u64);
+
+    format!("{}{stdd:?}", if d.lt(&Duration::zero()) { "-" } else { "" })
+}
+
+impl PacketGroup {
+    fn new() -> Self {
```

Comment:
Or just `#[derive(Default)]`?

Reply:
Neede to implement manually as chrono::Duration doesn't implement it, but yes it is nicer that way :-)